### PR TITLE
ODATA-1583

### PR DESCRIPTION
**Pull request by ralfhandl**

### CODE REVIEW

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

File: abnf/odata-abnf-construction-rules.txt [hunk 1]
```diff
@@ -429,13 +442,30 @@ contextFragment = %s"Collection($ref)"
                 / singletonEntity [ navigation *( containmentNavigation ) [ "/" qualifiedEntityTypeName ] ] [ selectList ]
                 / qualifiedTypeName [ selectList ]
                 / entitySet ( %s"/$deletedEntity" / %s"/$link" / %s"/$deletedLink" )
-                / entitySet keyPredicate "/" contextPropertyPath [ selectList ]
+                / entitySet contextKeyPredicate "/" contextPropertyPath [ selectList ]
```
Comment (HeikoTheissen) [inline abnf/odata-abnf-construction-rules.txt:445]: That still does not rule out percent-encoding in the `selectList`:
```
"@context": "Addresses(City%2CStreet)"
```

File: abnf/odata-abnf-construction-rules.txt [hunk 2]
```diff
@@ -429,13 +442,30 @@ contextFragment = %s"Collection($ref)"
...
+                 / dateTimeOffsetValue ; TODO: no percent-encoding in SIGN
+                 / dateValue
+                 / timeOfDayValue
+                 / decimalValue        ; TODO: no percent-encoding in SIGN
+                 / sbyteValue          ; TODO: no percent-encoding in SIGN
+                 / byteValue
+                 / int16Value          ; TODO: no percent-encoding in SIGN
+                 / int32Value          ; TODO: no percent-encoding in SIGN
+                 / int64Value          ; TODO: no percent-encoding in SIGN
+                 / string              ; TODO: no percent-encoding
```
[the remark this answers is not in the document]
Comment (ralfhandl) [reply abnf/odata-abnf-construction-rules.txt:466]: Yes, rule `string` currently is for URL-encoded strings, and there `%` is only allowed for percent-encoding, followed by two hex digits.

I'll wait for the TC decision on percent-encoding of fragments in context URLs before going into the tedious details.

File: abnf/odata-abnf-construction-rules.txt [hunk 3]
```diff
@@ -429,16 +442,34 @@ contextFragment = %s"Collection($ref)"
...
+                 / decimalValue
+                 / sbyteValue
+                 / byte
+                 / int16Value
+                 / int32Value
+                 / int64Value
+                 / stringValue
+                 / durationValue
+                 / enumValue
+stringValue = %x27 *( %x00-26 / %x27 %x27 / %x28-FF) %x27 ; SQUOTE = %x27
```
[the remark this answers is not in the document]
Comment (ralfhandl) [reply abnf/odata-abnf-construction-rules.txt:469]: If you can find the rule, just insert it.

**Review by HeikoTheissen — APPROVED** [review]

**Review by HeikoTheissen — APPROVED** [review]

File: abnf/odata-abnf-construction-rules.txt [hunk 4]
```diff
@@ -442,16 +442,34 @@ contextFragment = %s"Collection($ref)"
...
```
Comment (HeikoTheissen) [inline abnf/odata-abnf-construction-rules.txt:469]:
```suggestion
stringValue = %x27 *( %x00-26 / %x27 %x27 / %x28-2E / "%2f" / %x30-FF) %x27 ; SQUOTE = %x27
```
According to TC 2024-02-21, `/` must be encoded in a stringValue.

Comment (ralfhandl) [reply abnf/odata-abnf-construction-rules.txt:469]: That was discussed as a possibility, no decision yet.

Comment (HeikoTheissen) [reply abnf/odata-abnf-construction-rules.txt:469]: TC 2024-05-15 agrees not to encode these characters.

File: abnf/odata-abnf-testcases.yaml [hunk 5]
```diff
@@ -3463,10 +3479,22 @@ TestCases:
     Rule: context
     Input: "#Customers('ALFKI')/Orders"
 
+  - Name: Context URL - collection with containment, key value without percent-encoding
+    Rule: context
+    Input: "#Customers('ALF/KI')/Orders"
```
Comment (ralfhandl) [inline abnf/odata-abnf-testcases.yaml:3484]: Add negative test case for this in key-as-segment notation

Comment (ralfhandl) [reply abnf/odata-abnf-testcases.yaml:3489]: Already exists, line 3478

**Review by mikepizzo — APPROVED** [review]

**Pull request merged**